doc/designs/hsm.md

```diff
@@ -0,0 +1,277 @@
+
+Access to the HSM is done using a PKCS#11-compliant shared library.
+
+The mechanism for sharing access to the key data for replicas/clones is HSM-specific and generally as a network accessible
+device.
+
+## Use Cases
+
+- Install IPA with a CA which has its private keys stored on an HSM self-signed
+- Install IPA with a CA which has its private keys stored on an HSM externally signed
+- Install an IPA replica CA which utilizes the same HSM-stored private keys
+- Install a KRA with private keys stored in an HSM
+- Install a KRA clone which utilizes the same HSM-stored private keys
+
+Installing with an HSM is all or nothing. There will be no mixing
+and matching the subsystems stored in the HSM. For example, one will
+not be able to install the KRA on an HSM without the CA as well.
+
+Only RSA keys will be supported.
+
+
+## How to Use
+
+Using an HSM should be largely invisible to users and administrators beyond passing additional options during installation. The options required and any pre-installion work are HSM-specific.
+
+It will not be possible to mix and match by default. PKI supports specifying the token values so a user can override these using --pki-config-override but it is, and will be, untested.
+
+### Installation
+
+#### CA
+
+The token name, module name and shared library must be provided to the
+CA and KRA installers. These define where the keys will be stored
+and the naming convention for them.
+
+The CA signing, OCSP, Audit and Subsystem private keys and certificates
+are generated and stored in the HSM.
+
+| Option | Description                                       |
+|:------------------------ | :------------------------------ |
+| --token-name             | NSS name for the token          |
+| --library-path           | Path to PKCS#11 shared library  |
+| --token-password         | Password for the token          |
+
+This information will be stored in new schema so that replicas can auto-detect when an HSM is configured.
+
+ipa-ca-install will accept the same options.
+
+```
+attributeTypes: (
+  2.16.840.1.113730.3.8.21.1.TBD
+  NAME 'ipaCaHSMConfiguration'
+  DESC 'HSM Configuration'
+  SYNTAX 1.3.6.1.4.1.1466.115.121.1.15
+  SINGLE-VALUE
+  X-ORIGIN 'IPA v4.10'
+)
+```
+This will be added as a MAY to the ```ipaCa``` objectclass.
+
+This attribute will be semi-colon delimited and contain the HSM information needed to install a replica in the form of:
+
+token-name;library-path
+
+The token password will be prompted by ipa-replica-install or passed on the cli.
+
+The presence of this attribute is enough to indicate that an HSM is present. If components are missing then the user must provide them on the cli. When ever a replica, replica CA, KRA or replica KRA is added this attribute will be examined to determine whether an HSM is available or not, and what the options are. The options will be populated if available.
+
+token-name and library-path can be overridden by passing them on the command-line.
```

Comment:
Fair point. I think that library-path should have an option since it may not be located in the same place on all distributions, or architectures. I think a different token name means a different token and they cannot be mixed.